File: csharp/uwapi/commands.cs
using System.Runtime.InteropServices;

namespace Unnatural
{
    using Order = Interop.UwOrder;
    using Type = Interop.UwOrderTypeEnum;
    using Priority = Interop.UwOrderPriorityFlags;

    public static class Commands
    {
        public const uint Invalid = uint.MaxValue;

        public static Order[] Orders(uint unitId)
        {
            Interop.UwOrders os = new Interop.UwOrders();
            Interop.uwOrders(unitId, ref os);
            Order[] tmp = new Order[os.count];
            for (int i = 0; i < os.count; i++)
                tmp[i] = Marshal.PtrToStructure<Order>(os.orders + i * Marshal.SizeOf<Order>());
            return tmp;
        }

        public static void Order(uint unitId, Order order)
        {
            Interop.uwOrder(unitId, ref order);
        }

        public static Order Stop()
        {
            Order o = new Order();
            o.entity = Invalid;
            o.position = Invalid;
            o.order = Type.Stop;
            o.priority = Priority.User;
            return o;
        }

        public static Order Guard()
        {
            Order o = new Order();
            o.entity = Invalid;
            o.position = Invalid;
            o.order = Type.Guard;
            o.priority = Priority.User;
            return o;
        }

        public static Order RunToPosition(uint position)
        {
            Order o = new Order();
            o.entity = Invalid;
            o.position = position;
            o.order = Type.Run;
            o.priority = Priority.User;
            return o;
        }

        public static Order RunToEntity(uint entityId)
        {
            Order o = new Order();
            o.entity = entityId;
            o.position = Invalid;
            o.order = Type.Run;
            o.priority = Priority.User;
            return o;
        }

        public static Order FightToPosition(uint position)
        {
            Order o = new Order();
            o.entity = Invalid;
            o.position = position;
            o.order = Type.Fight;
            o.priority = Priority.User;
            return o;
        }

        public static Order FightToEntity(uint entityId)
        {
            Order o = new Order();
            o.entity = entityId;
            o.position = Invalid;
            o.order = Type.Fight;
            o.priority = Priority.User;
            return o;
        }

        public static void PlaceConstruction(uint constructionProto, uint position, float yaw = 0, uint recipeProto = 0, Interop.UwPriorityEnum priority = Interop.UwPriorityEnum.Normal)
        {
            Interop.uwCommandPlaceConstruction(constructionProto, position, yaw, recipeProto, priority);
        }

        public static void SetRecipe(uint unitId, uint recipeProto)
        {
            Interop.uwCommandSetRecipe(unitId, recipeProto);
        }

        public static void SetPriority(uint unitId, Interop.UwPriorityEnum priority)
        {
            Interop.uwCommandSetPriority(unitId, priority);
        }

        public static void Load(uint unitId, uint resourceProto)
        {
            Interop.uwCommandLoad(unitId, resourceProto);
        }

        public static void Unload(uint unitId)
        {
            Interop.uwCommandUnload(unitId);
        }

        public static void Move(uint unitId, uint position, float yaw = 0)
        {
            Interop.uwCommandMove(unitId, position, yaw);
        }

        public static void Aim(uint unitId, uint targetId)
        {
            Interop.uwCommandAim(unitId, targetId);
        }

        public static void RenounceControl(uint entityId)
        {
            Interop.uwCommandRenounceControl(entityId);
        }

        public static void SelfDestruct(uint entityId)
        {
            Interop.uwCommandSelfDestruct(entityId);
        }
    }
}

File: python/uwapi/commands.py
from .interop import *

class Commands:
    _instance = None

    def __new__(cls):
        if cls._instance is None:
            cls._instance = super().__new__(cls)
        return cls._instance

    def orders(self, unit_id: int) -> list[UwOrder]:
        return uw_interop.uwOrders(unit_id).orders

    def order(self, unit_id: int, order: UwOrder) -> None:
        uw_interop.uwOrder(unit_id, order)

    def stop(self) -> UwOrder:
        o = self._default_order()
        o.order = UwOrderTypeEnum.Stop
        return o

    def guard(self) -> UwOrder:
        o = self._default_order()
        o.order = UwOrderTypeEnum.Guard
        return o

    def run_to_position(self, position: int) -> UwOrder:
        o = self._default_order()
        o.position = position
        o.order = UwOrderTypeEnum.Run
        return o

    def run_to_entity(self, entity_id: int) -> UwOrder:
        o = self._default_order()
        o.entity = entity_id
        o.order = UwOrderTypeEnum.Run
        return o

    def fight_to_position(self, position: int) -> UwOrder:
        o = self._default_order()
        o.position = position
        o.order = UwOrderTypeEnum.Fight
        return o

    def fight_to_entity(self, entity_id: int) -> UwOrder:
        o = self._default_order()
        o.entity = entity_id
        o.order = UwOrderTypeEnum.Fight
        return o

    def place_construction(
        self,
        construction_proto: int,
        position: int,
        yaw: float = 0,
        recipe_proto: int = 0,
        priority: UwPriorityEnum = UwPriorityEnum.Normal
    ) -> None:
        uw_interop.uwCommandPlaceConstruction(construction_proto, position, yaw, recipe_proto, priority)

    def set_recipe(self, unit_id: int, recipe_proto: int) -> None:
        uw_interop.uwCommandSetRecipe(unit_id, recipe_proto)

    def set_priority(self, unit_id: int, priority : UwPriorityEnum) -> None:
        uw_interop.uwCommandSetPriority(unit_id, priority)

    def load(self, unit_id: int, resource_proto: int) -> None:
        uw_interop.uwCommandLoad(unit_id, resource_proto)

    def unload(self, unit_id: int) -> None:
        uw_interop.uwCommandUnload(unit_id)

    def move(self, unit_id: int, position: int, yaw: float = 0) -> None:
        uw_interop.uwCommandMove(unit_id, position, yaw)

    def aim(self, unit_id: int, target_id: int) -> None:
        uw_interop.uwCommandAim(unit_id, target_id)

    def renounce_control(self, entity_id: int) -> None:
        uw_interop.uwCommandRenounceControl(entity_id)

    def self_destruct(self, entity_id: int) -> None:
        uw_interop.uwCommandSelfDestruct(entity_id)

    def _default_order(self) -> UwOrder:
        return UwOrder(INVALID, INVALID, UwOrderTypeEnum.Nothing, UwOrderPriorityFlags.User)

uw_commands = Commands()
